src/lib/MerkleTree.ts

```diff
@@ -0,0 +1,219 @@
+      let subtree = this.subtrees[i];
+      if (subtree) {
+        // If there is already a smaller subtree, merge them.
+        if (smallestSubtree) {
+          const parent: MerkleNode = {
+            hash: Buffer.concat([subtree.hash, smallestSubtree.hash]),
+            firstChild: subtree,
+            secondChild: smallestSubtree,
+            parent: undefined
+          };
+
+          subtree.parent = parent;
+          smallestSubtree.parent = parent;
+
+          // The parent becomes the new smallest subtree.
+          smallestSubtree = parent;
+        } else { // There isn't already a smaller subtree, assign subtree as smallest.
+          smallestSubtree = this.subtrees[i];
+        }
+
+        this.subtrees[i] = undefined;
+      }
+    }
+
+    this.merkleTreeRootNode = smallestSubtree;
+
+    if (!this.merkleTreeRootNode) {
+      throw new Error('No value(s) given to construct a Merkle tree.');
+    }
+
+    return this.merkleTreeRootNode.hash;
+  }
+
+  /**
+   * Linearize the Merkle tree as a list of values.
+   * The output can be fed into initialize(...) to reconstruct the same Merkle tree.
+   * This method can also be used to perform custom serialization.
+   */
+  public linearize (): Buffer[] {
+    const values = Array.from(this.valueToMerkleNodeMap.keys());
+    return values;
+  }
+
+  /**
+   * Serailizes the Merkle tree.
+   */
+  public serialize (): Buffer {
+    const values: string[] = [];
+
+    for (let value of this.linearize()) {
+      // TODO: revisit choice of encoding.
+      const base64Value = value.toString('base64');
+      values.push(base64Value);
+    }
+
+    const buffer = Buffer.from(JSON.stringify(values));
+    return buffer;
+  }
+
+  /**
+   * Deserailizes the Merkle tree that was serialized using the serialize() function.
+   */
+  public static deserialize (valuesBuffer: Buffer): MerkleTree {
+    const valuesBase64: string[] = JSON.parse(valuesBuffer.toString());
+    const values: Buffer[] = [];
+
+    for (let valueBase64 of valuesBase64) {
+      values.push(Buffer.from(valueBase64, 'base64'));
+    }
+
+    const merkleTree = MerkleTree.initialize(values);
+    merkleTree.finalize();
+
+    return merkleTree;
+  }
+
+  /**
+   * Create a Merkle receipt for the given value.
+   */
+  public receipt (_value: Buffer): MerkleReceipt {
+    throw new Error('Not implemented.');
+  }
+
+  /**
+   * Proves that the given receipt is valid for the value given.
+   */
+  public static prove (_value: Buffer, _receipt: MerkleReceipt): boolean {
+    throw new Error('Not implemented.');
+  }
+
+}
+
+/**
+ * Internal representation of a node in a Merkle tree.
+ */
+interface MerkleNode {
+  hash: Buffer;
+  parent?: MerkleNode;
+  firstChild?: MerkleNode;
+  secondChild?: MerkleNode;
+}
+
+/**
+ * Class containing a Merkle receipt structure and its related operations.
+ */
+class MerkleReceipt {
+}
+
+/**
+ * SHA256 hash function.
+ */
+function sha256hash (value: Buffer): Buffer {
```

Comment:
Shouldn't this be changed to a const crypto.createHash('sha256') rather the creating on every invocation?

Reply:
No, you can create only one digest/hash per instance of Crypto.Hash created by createHash(…).